source/create.ts

```diff
@@ -26,15 +27,17 @@ export type HTTPAlias =
 	| 'head'
 	| 'delete';
 
-export type ReturnResponse = (url: URLArgument | Options & {stream?: false; url: URLArgument}, options?: Options & {stream?: false}) => CancelableRequest<Response>;
-export type ReturnStream = (url: URLArgument | Options & {stream: true; url: URLArgument}, options?: Options & {stream: true}) => ProxyStream;
-export type GotReturn = ProxyStream | CancelableRequest<Response>;
+export type ReturnResponse = (url: URLOrOptions | Options & {stream?: false}, options?: Options & {stream?: false}) => ReturnType<typeof asPromise>;
+export type ReturnStream = (url: URLOrOptions | Options & {stream?: true}, options?: Options & {stream?: true}) => ReturnType<typeof asStream>;
+export type GotReturn = ReturnType<ReturnResponse> | ReturnType<ReturnStream>;
```

Comment:
:+1:

Reply:
Actually, this can be further slimmed down in the instance type for `Got` below. One could use `Parameters<T>` and `ResponseType<T>`, but I am not sure if it will hurt readability in the users' side.